---
title: "Download NCBI fasta"
output: html_notebook
params:
  working_dir: /home/rstudio/projects
  data_dir: data
  probes_dir: probes/probes_May_27_2020
---

Download the Sars-cov2 fasta file from NCBI using E-utils. (Unfortunately E-utils is separate from the NCBI virus website data.  The virus website does not support E-utils.)

This URL(from the virus website): https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/virus?SeqType_s=Nucleotide&VirusLineage_ss=SARS-CoV-2,%20taxid:2697049 
returns a different number of sequences than the NCBI E-utils resource:
https://www.ncbi.nlm.nih.gov/nuccore/?term=sars-cov2%5BORGN%5D+OR+2697049%5BTXID%5D

The sequences from the Virus site eventually end up in the NCBI e-utils site but it might take a few days for the sequences to be indexed there. 

```{r required_libraries, message=FALSE, warning=FALSE}
if (! requireNamespace("rentrez", quietly=TRUE)) {
  install.packages("rentrez")
}
if (! requireNamespace("XML", quietly=TRUE)) {
  install.packages("XML")
}
if (! requireNamespace("ggplot2", quietly=TRUE)) {
  install.packages("ggplot2")
}

library(rentrez)
library(XML)
library(ggplot2)
library(knitr)

working_dir <- file.path(params$working_dir)
```


```{r include=FALSE}
#initialize directories
data_directory <- file.path(working_dir,params$data_dir)
if(!dir.exists(data_directory)){
  dir.create(data_directory)
}

fasta_directory <- file.path(data_directory,"fasta")
if(!dir.exists(fasta_directory )){
  dir.create(fasta_directory )
}

figures_directory <- file.path(working_dir,"figures")
if(!dir.exists(figures_directory)){
  dir.create(figures_directory )
}
```


## Methods

### Probe sequences
Several probes targeting each of ORF1ab, RdRp and E genes of the SARS-CoV-2 genome were provided by the Chan lab (Table 1).

```{r output_current_probes, include=FALSE}

probes_dir <- file.path(data_directory,params$probes_dir)

probe_files <- sort(list.files(probes_dir))

#from the file name get the gene and the type of tag
probes <- data.frame(name=character(),gene=character(),
                     tag_type=character(),
                     #sequence = character(),
                     stringsAsFactors = FALSE)
for(i in 1:length(probe_files)){
  current_file <- gsub(probe_files[i],pattern = ".fa",replacement = "")

  name <- current_file
  gene <- unlist(strsplit(current_file,split = "_"))[2]
  tag_type <- unlist(strsplit(current_file,split = "_"))[3]
  #sequence <- read.table(file.path(probes_dir,probe_files[i]),stringsAsFactors = FALSE)[2,1]
  probes <- rbind(probes, cbind(name, gene, tag_type))
  #probes <- rbind(probes, cbind(name, gene, tag_type,sequence))
}

kable(probes)
```



### Fasta Sequences
Download the fasta format as well as the meta information for all the sars-cov2 genomes available through  NCBI e-utils.

```{r download_ncbi_fasta, message=FALSE, warning=FALSE, include=FALSE}
#get all the sequences at NCBI for SARs-cov2
sequence_summary <- entrez_search(db = "nuccore", 
                                  "sars-cov2[ORGN] OR 2697049[TAXID]", 
                                  use_history = TRUE)

num_sequences <- sequence_summary$count

#get all the sequences as fasta and create one fasta file.
fname <- file.path(fasta_directory ,paste("ncbi_sars_cov2_", Sys.Date(), ".fa", sep = ""))
meta_data <- c()
for (start_rec in seq(0, num_sequences, 50)) {

  recs <- rentrez::entrez_fetch(db = "nuccore", 
                       web_history = sequence_summary$web_history ,  
                       retstart = start_rec, retmax = 50, rettype="fasta")
  xml <- rentrez::entrez_fetch(db = "nuccore", 
                      web_history = sequence_summary$web_history ,  
                       retstart = start_rec, retmax = 50, rettype="gb",
                      retmode = "xml")
  
  #convert the xml to dataframe
  xml_converted <- XML::xmlToDataFrame(xml,stringsAsFactors = FALSE)
  
  #add to metadata dataframe
  if(length(meta_data) == 0){
    meta_data <- xml_converted
  } else{
    if(length(colnames(meta_data)) != length(colnames(xml_converted))){
      missing_colnames_meta <- setdiff(colnames(xml_converted), 
                                       colnames(meta_data))
      missing_colnames_xml <- setdiff(colnames(meta_data), 
                                      colnames(xml_converted))
      if(length(missing_colnames_meta) >0){
        for(i in 1:length(missing_colnames_meta)){
          meta_data[,length(colnames(meta_data))+1] <- NA
          colnames(meta_data)[ncol(meta_data)] <- missing_colnames_meta[i]
        }
      }
      if(length(missing_colnames_xml) >0){
        for(i in 1:length(missing_colnames_xml)){
          xml_converted[,length(colnames(xml_converted))+1] <- NA
          colnames(xml_converted)[ncol(xml_converted)] <- missing_colnames_xml[i]
        }
      }
    }
    
    meta_data <- rbind(meta_data, xml_converted)
  }
  write(recs, fname,append = TRUE)
  print(paste("wrote records to ", fname, sep = ""))
}
```

NCBI is an established nucleotide repository and serves as a second source of global SARS-CoV-2 genomes. `r dim(meta_data)[1]` sequences (selection criteria, nucleotide completeness: ‘complete’) were downloaded from the NCBI Virus Hub (https://www.ncbi.nlm.nih.gov/labs/virus/vssi) on `r Sys.Date() ` . The sequences include isolates from multiple sources (Fig 3).


```{r echo=FALSE}

#create a plot of data growth
dat <- meta_data
dt <- as.Date(meta_data$`GBSeq_update-date`,format = "%d-%B-%y")
dat$yr <- as.integer(substr(dt,1,4))
dat$mon <- as.integer(substr(dt,6,7))
dat$day <- as.integer(substr(dt,9,10))

idx <- order(dat$yr,dat$mon,dat$day)
dat <- dat[idx,]
dat$date2 <- sprintf("%s-%s-%s",dat$yr,dat$mon,dat$day)
uq <- dat$date2[!duplicated(dat$date2)]
dat$date2 <- factor(dat$date2,ordered=TRUE,labels=uq)

dt <- format(Sys.Date(),"%y%m%d")

ncbi_growth_figure <- file.path(figures_directory,sprintf("NCBI_count_%s.png",dt))

png(ncbi_growth_figure,width=13,height=6)
tryCatch({
  p <- ggplot(dat, aes(date2))
  p <- p + geom_bar() + xlab("Release date") + ylab("Num genomes")
  p <- p + ggtitle("NCBI SARS-CoV-2 genomes: Deposition over time")
  p <- p + theme(axis.text.x = element_text(angle = 90, hjust = 1,size=8),
                 axis.text.y=element_text(size=15))
  print(p)
},error=function(ex) {
  print(ex)
},finally={
  dev.off()
})
p

```
