# BLAST Pairwise Sequence Alignment 
run_blast.sh

# Description
This script runs blastn on an input query.fa (probe) against n subject sequences (viral genomes).

## Input
 * probe name (i.e. Egene)
 * probe sequence (i.e. CGTTAATAGTTAATAGCGTACTTCTTTTTC)
 * type (i.e. reporter, capture, forward, reverse, reporter2)

## Output
 * a file called blast_out_<PROBE_NAME>_<TYPE>.txt  with blastn staitistics and info (i.e. qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore)
 * probe_<PROBE_NAME>_<TYPE>.fa (can be deleted - autogenerated probe file in fasta format)

## Usage
./run_blast.sh <PROBE_NAME> <PROBE_SEQ> <TYPE>

```
./run_blast.sh Egene CGTTAATAGTTAATAGCGTACTTCTTTTTC forward
```
